### ISSUE TITLE
Feature request: Implement another shm plugin in POSIX Shared Memory

### ISSUE THREAD

**Issue by Torrekie** [TigerVNC/tigervnc#1804]
Since some os does not have builtin SysV-shm (e.g. non-macOS Darwin), implementations though POSIX shared memory is preferred.

Since I'm working on a Xfce4 Darwin port, I've noticed some limitations of system calls.

- `<sys/shm.h>`, `<sys/sem.h>` is completely broken since Apple did not compile the SysV features into iOS kernel.
- `mmap()` is an alternative method which used by current xserver, but it was strictly forbidden by sandbox, which means it can only be executed out-of-sandbox.
- `shm_open()` which is POSIX-compliant, but does not support listing like how SysV `ipcs(1)` does, the only way to remove opened shm is to use `shm_unlink()` with correct name, or reboot the device.

The `mmap()` method is useable but sometimes limited since the homedir of default users (`root` and `mobile`) is sandboxed, executing under homedir could be rejected by dyld and produce error `file system sandbox blocked mmap()`. So `shm_open()` is still considered a better solution.

**Comment by CendioOssman** [comment]
I'm afraid I don't quite understand your setup. Do you want to run vncviewer, in Xfce, that's using XQuartz? On iOS?

**Comment by Torrekie** [comment]
there's an [article](https://maxleiter.com/blog/X11) talked about entering X11 window though vncserver that running nativity on iOS, and what I'm doing is basically same. The problem is some of server codes are using SysV shm which is not really available on iOS (and possibly other non-SysV systems).

What I wanted to say is that, can we *use SysV alternatives* that based on maybe POSIX shm or whatever which makes the program behaves basically same as how it works on systems that supported SysV.

**Comment by CendioOssman** [comment]
Which code specifically are you talking about? I think the server code is all in Xorg. I.e., it's not part of the TigerVNC project.

**Comment by Torrekie** [comment]
I'm sure `grep -E "HAS_SHM|MITSHM|shmctl|shmdt|shmget|shmop" -rl` are all codes that using SysV shm in this project, `HAS_SHM` and `MITSHM` are defined in `config.h` to control whether MIT-SHM plugin to be built.

**Comment by CendioOssman** [comment]
Yes, but those are for the client. From that link, it sounds like you are interested in the server?

**Comment by Torrekie** [comment]
i wish both client and server can keep desired behavior after patching, I didn't managed to read through the whole project so I can't make sure if it can work as expected by simply replacing the shm implementation

**Comment by CendioOssman** [comment]
I'm afraid that iOS isn't really something we're targeting at the moment. And from what you've described so far, the main work would need to be done in Xorg, not in TigerVNC.

Please raise the issue with them first. Once that is in place, we can see what's left to do in TigerVNC.
